src/twisted/web/http_headers.py

```diff
@@ -209,23 +218,50 @@ def setRawHeaders(self, name, values):
                 "instance of %r instead" % (name, type(values))
             )
 
+        if not isinstance(name, (bytes, str)):
+            raise TypeError(
+                "Header name is an instance of %r, " "not bytes or str" % (type(name),)
+            )
+
+        for count, value in enumerate(values):
+            if not isinstance(value, (bytes, str)):
+                raise TypeError(
+                    "Header value at position %s is an instance of %r, not "
+                    "bytes or str"
+                    % (
+                        count,
+                        type(value),
+                    )
+                )
+
         name = _sanitizeLinearWhitespace(self._encodeName(name))
         encodedValues = [
             _sanitizeLinearWhitespace(v) for v in self._encodeValues(values)
         ]
 
-        self._rawHeaders[name] = self._encodeValues(encodedValues)
+        self._rawHeaders[name] = encodedValues
 
-    def addRawHeader(self, name, value):
+    def addRawHeader(self, name: Union[bytes, str], value: Union[bytes, str]) -> None:
         """
         Add a new raw value for the given header.
 
-        @type name: L{bytes} or L{unicode}
+        @type name: L{bytes} or L{str}
         @param name: The name of the header for which to set the value.
 
-        @type value: L{bytes} or L{unicode}
+        @type value: L{bytes} or L{str}
         @param value: The value to set for the named header.
         """
+        if not isinstance(name, (bytes, str)):
+            raise TypeError(
+                "Header name is an instance of %r, " "not bytes or str" % (type(name),)
+            )
+
+        if not isinstance(value, (bytes, str)):
+            raise TypeError(
+                "Header value is an instance of %r, not "
+                "bytes or str" % (type(value),)
+            )
+
```

Comment:
Same as above, why in this PR?

Reply:
This came from @hawkowl 's original PR here: https://github.com/twisted/twisted/pull/1284/files#diff-effe5888987605608f990d71470cc65dR223